### DISCUSSION TITLE
Does PlanetScale support the .NET 6/7 EF Code First and reverse engineering?

### DISCUSSION THREAD

**Discussion by SurferOnWww-new** [planetscale/discussion#391] [Q&A]
As far as I have tested in my environment (Windows 10.0.19045, Visual Studio 2022 17.4.4, .NET SDKs 6.0.308 / 7.0.102, .NET runtimes 6.0.13 / 7.0.2) it does not.

(1) EF Code First

Add-Migration seems to be OK but Update-Database does not work.

MySql.EntityFrameworkCore 6.0.10 and 7.0.0 result in KeyNotFoundException.

Pomelo.EntityFrameworkCore.MySql 6.0.2 and 7.0.0 result in failure of ALTER DATABASE CHARACTER SET utf8mb4 and the error message "alter database is not supported" is shown.

(2) Reverse Engineering

The reverse engineering is the process to create the context class and entity classes based on the existing database. See the following document for details:

[Reverse Engineering](https://learn.microsoft.com/en-us/ef/core/managing-schemas/scaffolding/?tabs=dotnet-core-cli)

MySql.EntityFrameworkCore 6.0.10 and 7.0.0 does not work at all due to KeyNotFoundException.

Pomelo.EntityFrameworkCore.MySql 6.0.2 and 7.0.0 can create the context class only. However it is not complete.

Does the current PlanetScale support the .NET EF Code First and reverse engineering? If not is there a plan to support them in the future?

**Comment by orware** [comment]
Hello @SurferOnWww-new,

Please excuse my lack of direct familiarity with the options available for .NET, as my knowledge currently extends only to what I've read so far related to the recent [discussions#390](https://github.com/planetscale/discussion/discussions/390) topic from last Friday. 

I am planning on running some tests to gather more data on this topic related to these two recent discussions so I can understand the issues you are experiencing a little better, but any extra information you can help provide will be useful as this may be a complicated enough situation where more digging is needed to understand the issue and if there is anything we might be able to do to get things working.

For the EF Code First Topic, when using `MySql.EntityFrameworkCore`, do you still receive the same `KeyNotFoundException` error even when attempting to use something similar to what's described in the following "[Code First to an Existing Database](https://learn.microsoft.com/en-us/ef/ef6/modeling/code-first/workflows/existing-database)" documentation? This may have been what you were already describing above, but I would like to confirm.

The issue with `Pomelo.EntityFrameworkCore.MySql` is an interesting one. In this situation, does it always attempt an `ALTER DATABASE CHARACTER SET utf8mb4` operation and there is no way to avoid/prevent that from occurring? An error like this would be expected since the database would already be created for your PlanetScale database and those types of `ALTER` statements at the database level aren't allowed, so we would want these frameworks to have an option of not attempting those if possible.

For the Reverse Engineering Topic, when using the `Pomelo.EntityFrameworkCore.MySql` option, is that also using EF only, or is the MySQLConnector option available for that not the same as using EF?

This seems like something we would want to ensure is working correctly for any .NET developers wanting to use PlanetScale databases, but it'll take a bit of extra research / testing to get there if there are issues currently with the available EF options so any extra details you can share will be helpful as well as allowing for a bit of time to get some tests completed and come back with any updates I can provide.

**Reply by SurferOnWww-new** [reply]
Thank you for reply, @orware.

> For the EF Code First Topic, when using MySql.EntityFrameworkCore, do you still receive the same KeyNotFoundException error even when attempting to use something similar to what's described in the following "Code First to an Existing Database" documentation?

Yes. I have tested the EF Code First operation according to the following Microsoft tutorial:

[Part 4, add a model to an ASP.NET Core MVC app](https://learn.microsoft.com/en-us/aspnet/core/tutorials/first-mvc-app/adding-model?view=aspnetcore-7.0&tabs=visual-studio)

When I used the MySQL 8.0.19 installed in my local PC and MySql.EntityFrameworkCore the database and table are created as expected.

However, when I use the PlanetSacle the Update-Database command results in the KeyNotFoundException.

> The issue with Pomelo.EntityFrameworkCore.MySql is an interesting one. In this situation, does it always attempt an ALTER DATABASE CHARACTER SET utf8mb4 operation and there is no way to avoid/prevent that from occurring?

Below is the some detailed information when Pomelo.EntityFrameworkCore.MySql is used. Hope this will help.

(1) The NuGet packages used are:

Pomelo.EntityFrameworkCore.MySql 7.0.0
Microsoft.EntityFrameworkCore.Design 7.0.3
Microsoft.EntityFrameworkCore.SqlServer 7.0.3
Microsoft.EntityFrameworkCore.Tools 7.0.3
Microsoft.VisualStudio.Web.CodeGeneration.Design 7.0.4

(2) The context class created by scaffolding (the entity class is the same as the tutorial):

```
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Microsoft.EntityFrameworkCore;
using PlanetScaleMovie.Models;

namespace PlanetScaleMovie.Data
{
    public class PlanetScaleMovieContext : DbContext
    {
        public PlanetScaleMovieContext (DbContextOptions<PlanetScaleMovieContext> options)
            : base(options)
        {
        }

        public DbSet<PlanetScaleMovie.Models.NewMovie> NewMovie { get; set; } = default!;
    }
}
```

(3) 20230216015340_InitialCreate.cs created by Add-Migration InitialCreate command:

```
using System;
using Microsoft.EntityFrameworkCore.Metadata;
using Microsoft.EntityFrameworkCore.Migrations;

#nullable disable

namespace PlanetScaleMovie.Migrations
{
    /// <inheritdoc />
    public partial class InitialCreate : Migration
    {
        /// <inheritdoc />
        protected override void Up(MigrationBuilder migrationBuilder)
        {
            migrationBuilder.AlterDatabase()
                .Annotation("MySql:CharSet", "utf8mb4");

            migrationBuilder.CreateTable(
                name: "NewMovie",
                columns: table => new
                {
                    Id = table.Column<int>(type: "int", nullable: false)
                        .Annotation("MySql:ValueGenerationStrategy", MySqlValueGenerationStrategy.IdentityColumn),
                    Title = table.Column<string>(type: "longtext", nullable: true)
                        .Annotation("MySql:CharSet", "utf8mb4"),
                    ReleaseDate = table.Column<DateTime>(type: "datetime(6)", nullable: false),
                    Genre = table.Column<string>(type: "longtext", nullable: true)
                        .Annotation("MySql:CharSet", "utf8mb4"),
                    Price = table.Column<decimal>(type: "decimal(65,30)", nullable: false)
                },
                constraints: table =>
                {
                    table.PrimaryKey("PK_NewMovie", x => x.Id);
                })
                .Annotation("MySql:CharSet", "utf8mb4");
        }

        /// <inheritdoc />
        protected override void Down(MigrationBuilder migrationBuilder)
        {
            migrationBuilder.DropTable(
                name: "NewMovie");
        }
    }
}
```
(4) The result of Update-Database command:

PM> Update-Database
Build started...
Build succeeded.
Microsoft.EntityFrameworkCore.Database.Command[20101]
      Executed DbCommand (30ms) [Parameters=[], CommandType='Text', CommandTimeout='30']
      SELECT 1 FROM INFORMATION_SCHEMA.TABLES WHERE TABLE_SCHEMA='my-planetscale' AND TABLE_NAME='__EFMigrationsHistory';
Microsoft.EntityFrameworkCore.Database.Command[20101]
      Executed DbCommand (13ms) [Parameters=[], CommandType='Text', CommandTimeout='30']
      SELECT 1 FROM INFORMATION_SCHEMA.TABLES WHERE TABLE_SCHEMA='my-planetscale' AND TABLE_NAME='__EFMigrationsHistory';
Microsoft.EntityFrameworkCore.Database.Command[20101]
      Executed DbCommand (15ms) [Parameters=[], CommandType='Text', CommandTimeout='30']
      SELECT `MigrationId`, `ProductVersion`
      FROM `__EFMigrationsHistory`
      ORDER BY `MigrationId`;
Microsoft.EntityFrameworkCore.Migrations[20402]
      Applying migration '20230216015340_InitialCreate'.
Applying migration '20230216015340_InitialCreate'.
fail: Microsoft.EntityFrameworkCore.Database.Command[20102]
      Failed executing DbCommand (13ms) [Parameters=[], CommandType='Text', CommandTimeout='30']
      ALTER DATABASE CHARACTER SET utf8mb4;
Failed executing DbCommand (13ms) [Parameters=[], CommandType='Text', CommandTimeout='30']
ALTER DATABASE CHARACTER SET utf8mb4;
MySqlConnector.MySqlException (0x80004005): alter database is not supported
   at MySqlConnector.Core.ResultSet.ReadResultSetHeaderAsync(IOBehavior ioBehavior) in /_/src/MySqlConnector/Core/ResultSet.cs:line 43
   at MySqlConnector.MySqlDataReader.ActivateResultSet(CancellationToken cancellationToken) in /_/src/MySqlConnector/MySqlDataReader.cs:line 130
   at MySqlConnector.MySqlDataReader.CreateAsync(CommandListPosition commandListPosition, ICommandPayloadCreator payloadCreator, IDictionary`2 cachedProcedures, IMySqlCommand command, CommandBehavior behavior, Activity activity, IOBehavior ioBehavior, CancellationToken cancellationToken) in /_/src/MySqlConnector/MySqlDataReader.cs:line 468
   at MySqlConnector.Core.CommandExecutor.ExecuteReaderAsync(IReadOnlyList`1 commands, ICommandPayloadCreator payloadCreator, CommandBehavior behavior, Activity activity, IOBehavior ioBehavior, CancellationToken cancellationToken) in /_/src/MySqlConnector/Core/CommandExecutor.cs:line 56
   at MySqlConnector.MySqlCommand.ExecuteNonQueryAsync(IOBehavior ioBehavior, CancellationToken cancellationToken) in /_/src/MySqlConnector/MySqlCommand.cs:line 296
   at MySqlConnector.MySqlCommand.ExecuteNonQuery() in /_/src/MySqlConnector/MySqlCommand.cs:line 107
   at Microsoft.EntityFrameworkCore.Storage.RelationalCommand.ExecuteNonQuery(RelationalCommandParameterObject parameterObject)
   at Microsoft.EntityFrameworkCore.Migrations.MigrationCommand.ExecuteNonQuery(IRelationalConnection connection, IReadOnlyDictionary`2 parameterValues)
   at Microsoft.EntityFrameworkCore.Migrations.Internal.MigrationCommandExecutor.ExecuteNonQuery(IEnumerable`1 migrationCommands, IRelationalConnection connection)
   at Microsoft.EntityFrameworkCore.Migrations.Internal.Migrator.Migrate(String targetMigration)
   at Microsoft.EntityFrameworkCore.Design.Internal.MigrationsOperations.UpdateDatabase(String targetMigration, String connectionString, String contextType)
   at Microsoft.EntityFrameworkCore.Design.OperationExecutor.UpdateDatabaseImpl(String targetMigration, String connectionString, String contextType)
   at Microsoft.EntityFrameworkCore.Design.OperationExecutor.UpdateDatabase.<>c__DisplayClass0_0.<.ctor>b__0()
   at Microsoft.EntityFrameworkCore.Design.OperationExecutor.OperationBase.Execute(Action action)
alter database is not supported

**Reply by SurferOnWww-new** [reply]
> For the Reverse Engineering Topic, when using the Pomelo.EntityFrameworkCore.MySql option, is that also using EF only, or is the MySQLConnector option available for that not the same as using EF?

First, I created the movie table using the Console in the PlanetScale site as the EF Code First is not successful.

Then, I created a new .NET 6.0 console application, added the following NuGet packages and apply Scaffold-DbContext command on the PM console of Visual Studio to generate the context class and entity class:

Microsoft.EntityFrameworkCore.Tools 6.0.10
Pomelo.EntityFrameworkCore.MySql 6.0.2

Below is the generated context class. (Entity class was not generated at all)

```
using Microsoft.EntityFrameworkCore;
using Microsoft.EntityFrameworkCore.Metadata;
using ConsoleAppPlanetScale.Models;

namespace ConsoleAppPlanetScale.Data
{
    public partial class myplanetscaleContext : DbContext
    {
        public myplanetscaleContext()
        {
        }

        public myplanetscaleContext(DbContextOptions<myplanetscaleContext> options)
            : base(options)
        {
        }

        protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
        {
            if (!optionsBuilder.IsConfigured)
            {
#warning To protect potentially sensitive information in your connection string, you should move it out of source code. You can avoid scaffolding the connection string by using the Name= syntax to read it from configuration - see https://go.microsoft.com/fwlink/?linkid=2131148. For more guidance on storing connection strings, see http://go.microsoft.com/fwlink/?LinkId=723263.
                optionsBuilder.UseMySql("Server=ap-northeast.connect.psdb.cloud;Database=my-planetscale;user=5wbyfxpqblpyryeth41w;password=pscale_pw_xxxxx;SslMode=VerifyFull", Microsoft.EntityFrameworkCore.ServerVersion.Parse("8.0.23-mysql"));
            }
        }

        protected override void OnModelCreating(ModelBuilder modelBuilder)
        {
            modelBuilder.UseCollation("utf8mb4_0900_ai_ci")
                .HasCharSet("utf8mb4");

            OnModelCreatingPartial(modelBuilder);
        }

        partial void OnModelCreatingPartial(ModelBuilder modelBuilder);
    }
}
```

Please note that the above is not complete as the following code is not included:

```
public virtual DbSet<Movie> Movies { get; set; } = null!;
```

**Comment by kawpls** [comment]
is this solved?

**Reply by orware** [reply]
Hello @kawpls,

I haven't had an opportunity to test out the .NET options in any detail like our user above was able to share, so it would be dependent on whether they were able to find a working option for their needs above with a PlanetScale database.

I'm still not sure what the exact issue would be, but since there are differences on our end due to our extra Vitess layer that sits above MySQL 8 instances in our environment that would be a likely reason to explain the behavior difference compared to using these .NET options against a regular MySQL 8 database.

In [discussions#390](https://github.com/planetscale/discussion/discussions/390) there is an option there that seems to be able to get past the `KeyNotFoundException` issue, but I'm unsure if that would provide a full resolution to what @SurferOnWww-new was describing above.

**Comment by Josh865** [comment]
I was able to track down the cause of this issue to the line that calls `AlterDatabase()` on the first line of the `Up` method in the generated EF Core migration file:

```
public partial class InitialCreate : Migration
{
        /// <inheritdoc />
        protected override void Up(MigrationBuilder migrationBuilder)
        {
            migrationBuilder.AlterDatabase()
                .Annotation("MySql:CharSet", "utf8mb4");

            migrationBuilder.CreateTable(
                // ...
            )
        }
}
```

This makes sense, given that PlanetScale doesn't support `ALTER TABLE` as mentioned at https://planetscale.com/docs/reference/mysql-compatibility#queries-functions-syntax-data-types-and-sql-modes.

What worked for me is to disable Polemo's charset handling as discussed here: https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/wiki/Character-Sets-and-Collations#do-not-apply-any-charsets-and-collations. With that setting, the `AlterDatabase()` method is no longer called with each migration.

**Reply by orware** [reply]
Awesome work @Josh865!

The `ALTER TABLE` operations do work, but `ALTER DATABASE` wouldn't which is the main issue here. The page you linked to above mainly mentions that those operations can be destructive when it comes to renaming things primarily due to how the online DDL procedure functions on our end but we have the [handling table and column renames](https://planetscale.com/docs/learn/handling-table-and-column-renames) documentation page to help cover that topic separately.

The big detail you've connected above for myself and hopefully others though is the need to disable that option as you shared in the second link you provided.

**Reply by chrisleyva** [reply]
@Josh865 I'm not sure if I've done this right. I can't seem to get this to work.
When I run `Update-Database` in the PM Console, I get the following error: `VT12001: unsupported: ALTER DATABASE`

And I see that the `__EFMigrationsHistory` table was created.

Am I disabling the charset handling in the right place? (screenshot)
![image](https://github.com/planetscale/discussion/assets/1437143/3f2ca44c-adeb-45f8-8e8e-124c9b7d5ea3)

**Reply by Josh865** [reply]
That looks correct from what I can remember, but I haven't worked in the project I was using with PlanetScale in quite some time so I can't reference the code.

When I was exploring this problem, I had to delete all of the existing generated migration files and the entries in the `__EFMigrationsHistory` table between each attempt, otherwise the existing migrations (with the `AlterDatabase()` call) wouldn't be replaced. I'm not sure if that'a an option for you, but if so, it may be worth a try.

**Reply by chrisleyva** [reply]
Wow, you're quick! I'll give that a try. Thank you very much.

**Reply by chrisleyva** [reply]
Ok, I was able to successfully create the database by doing the following (based on [this link](https://stackoverflow.com/a/11679386/329928)):

```md
 1. Delete the state: Delete the migrations folder in your project
 2. Delete the `__EFMigrationHistory` table
 3. Run the following command in the Package Manager Console:
 
        Add-Migration Initial
```

**Comment by msh2050** [comment]
this is an easy (how to make it work)
1- use (Pomelo.EntityFrameworkCore.MySql) I did not test with (MySql.EntityFrameworkCore)
2- in the command line `dotnet ef migrations add InitialCreate`
3- delete or comment (on InitialCreate.cs migration file):
```c#
     migrationBuilder.AlterDatabase()
            .Annotation("MySql:CharSet", "utf8mb4");
```
4- to prevent CONSTRAINT error delete or comment any table.ForeignKey in (InitialCreate.cs) like:
```c#
table.ForeignKey(
                        name: "FK_mater_detail_detailId",
                        column: x => x.detailId,
                        principalTable: "details",
                        principalColumn: "Id");
```
5- in the command line  `dotnet ef database update`
6- this will not work on the main branch of your database create another branch and it will work

regards

**Reply by aldrin0408av** [reply]
How can I access the Navigational Properties?

**Reply by chrisleyva** [reply]
@msh2050  Does this solution mean that our schemas can't have foreignKeys at all?

**Reply by msh2050** [reply]
As in the PlanetScale [Documentation](https://planetscale.com/docs/learn/operating-without-foreign-key-constraints) it is clearly stated that PlanetScale doesn't support FOREIGN KEY constraints.
so you can use FOREIGN KEY but as it is written in the same documentation mentioned above `You are encouraged to use the relational model and associate tables by "pointing" rows from one table to another. It is just the enforcement at the database level, the CONSTRAINT ... FOREIGN KEY definition, that is not allowed in your schemas.
`
